ant-protocol/src/messages/merkle_tree.rs

```diff
@@ -0,0 +1,1821 @@
+            leaf_count,
+            depth,
+            salts,
+        })
+    }
+
+    /// Get the root hash of the tree
+    ///
+    /// This is the hash committed on-chain for batch payment
+    pub fn root(&self) -> XorName {
+        self.root
+    }
+
+    /// Get the depth of the tree
+    pub fn depth(&self) -> u8 {
+        self.depth
+    }
+
+    /// Get the original leaf count (before padding)
+    pub fn leaf_count(&self) -> usize {
+        self.leaf_count
+    }
+
+    /// Get midpoint nodes at depth/2
+    ///
+    /// These are the nodes used internally to determine candidate pools for payment routing.
+    /// Returns intersections nodes at level depth/2
+    ///
+    /// Note: Users typically don't need this directly - use `reward_candidates()` instead.
+    fn midpoints(&self) -> Result<Vec<MerkleMidpoint>> {
+        let midpoint_level = (self.depth / 2) as usize;
+
+        let nodes =
+            self.inner
+                .get_nodes_at_level(midpoint_level)
+                .ok_or(MerkleTreeError::Internal(
+                    "Midpoint level must exist".to_string(),
+                ))?;
+
+        let midpoints: Vec<MerkleMidpoint> = nodes
+            .into_iter()
+            .map(|(index, hash)| MerkleMidpoint {
+                hash: XorName(hash),
+                index,
+            })
+            .collect();
+
+        Ok(midpoints)
+    }
+
+    /// Get reward candidates for batch payment
+    ///
+    /// Computes candidate addresses as hash(midpoint_hash || root || tx_timestamp).
+    /// Network nodes closest to these addresses are eligible for payment rewards.
+    /// Each candidate contains a proof that the midpoint belongs to the tree.
+    /// Returns 2^ceil(depth/2) reward candidates.
+    ///
+    /// # Arguments
+    ///
+    /// * `tx_timestamp` - Unix timestamp of the transaction (seconds since epoch)
+    ///
+    /// # Returns
+    ///
+    /// A vector of `RewardCandidatePool` or an error if proof generation fails
+    ///
+    /// # Example
+    ///
+    /// ```ignore
+    /// let tree = MerkleTree::from_xornames(chunks)?;
+    /// let tx_timestamp = SystemTime::now()
+    ///     .duration_since(UNIX_EPOCH)
+    ///     .expect("Failed to get current time")
+    ///     .as_secs();
+    ///
+    /// let candidates = tree.reward_candidates(tx_timestamp)?;
+    ///
+    /// // Each candidate's branch can be verified independently
+    /// for candidate in candidates {
+    ///     assert!(candidate.branch.verify());
+    /// }
+    /// ```
+    pub fn reward_candidates(&self, tx_timestamp: u64) -> Result<Vec<RewardCandidatePool>> {
+        let midpoints = self.midpoints()?;
+
+        midpoints
+            .into_iter()
+            .map(|midpoint| {
+                // Generate proof for this midpoint
+                let branch = self.generate_midpoint_proof(midpoint.index, midpoint.hash)?;
+
+                Ok(RewardCandidatePool {
+                    branch,
```

Comment:
the `index` and `hash` of the midpoint better to be included in this struct entry as well ?